### PR TITLE
[#1] Payment flows cleanup

### LINKED ISSUE #1: Circle: Consolidate submission handlers & general cleanup
## Overview

- The checkout page has two large functions that are nearly identical (submit bid and submit purchase). 
  - Can we consolidate these into one function?
  - Both of these functions do an awful lot, can they be streamlined or split up?
- The add payment method page does something almost identical as well, so same thoughts here.
- Does it make sense to add a payment context? There's a good deal of prop drilling going on.

### PR DESCRIPTION
**Pull request by dallashuggins**
Closes #1 

### Changes
- New payment context
- Using new context methods in various payment flows (checkout, add card). 
- Abstracting logic used in previous methods to share between purchase and bid flows. 
- Ran `npm run format` and included some files I hadn't worked on here since the formatter found issues.

### CODE REVIEW

File: apps/web/src/contexts/payment-context.tsx [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+      securityCode: string,
+      cardId: string,
+      publicKeyRecord: PublicKey
+    ) => {
+      // Throw error if no price
+      if (!release || !release.price) {
+        throw new Error('No price provided')
+      }
+
+      // Throw error is no card is provided
```
Comment (dallashuggins) [inline apps/web/src/contexts/payment-context.tsx:156]:
```suggestion
      // Throw error if no card is provided
```

File: apps/web/src/contexts/payment-context.tsx [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+            encryptedData: encryptedCard,
+            saveCard: !!saveCard,
+            state,
+            zipCode,
+            default: saveCard ? !!defaultCard : false,
+          })
+          .catch(async (error) => {
+            const response = await error.response.json()
+            mapCircleErrors(response.code)
+            setFormErrors({}) // @TODO: Work with purchase?
```
Comment (dallashuggins) [inline apps/web/src/contexts/payment-context.tsx:288]:
```suggestion
            setFormErrors({})
```

**Review by afraser — APPROVED** [review]
Nice!

File: apps/web/src/contexts/payment-context.tsx [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+          | typeof validateFormExpirationDate
+        >
+      >
+    >()
+
+  const mapCircleErrors = useCallback(
+    (code: string | number) => {
+      setFormErrors({})
+
+      switch (code) {
```
Comment (afraser) [inline apps/web/src/contexts/payment-context.tsx:121]: pretty sure you could condense this to:

```
const errors = {
  1094: { fullName: t('forms:errors.invalidName') },
  1032: { ccNumber: t('forms:errors.invalidCreditCard') },
  2: { securityCode: t('forms:errors.invalidCVV') },
  1106: { state: t('forms:errors.invalidState') },
  1101: { country: t('forms:errors.invalidCountry') },
}[code]

if (errors) {
  return setFormErrors(errors)
}
```

Comment (dallashuggins) [reply apps/web/src/contexts/payment-context.tsx]: That does simplify it, thanks! Added in 52abf29

**Review by jakerainis — APPROVED** [review]
Looks great! So much clean!

**Review by smonn — APPROVED** [review]
Looks and works great, thanks!

**Pull request merged**